### PR TITLE
Touch #1416256: String tiers admin js

### PR DESCRIPTION
**Pull request by phlax**
Adds react widgets for string tiers admin ui

### CODE REVIEW

File: pontoon/static/js/tags/admin/errors.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        const {errors} = this.props;
+        if (!Object.keys(errors).length > 0) {
+            return '';
+        }
+        return (
+            <ul className="errors">
+                {Object.entries(errors).map(([name, error], key) => {
+                    return (
+                        <li className="error"
+                            style={{color: 'red'}}
```
Comment (phlax) [inline pontoon/static/js/tags/admin/errors.js]: remove css from here

File: pontoon/static/js/tags/admin/errors.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+
+import React from 'react';
+
+
+export default class TagResourceErrors extends React.Component {
```
Comment (phlax) [inline pontoon/static/js/tags/admin/errors.js]: wondering whether to move this to an `ErrorList` component

File: pontoon/static/js/tags/admin/table.js [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+               checked={checked.has(data.original[0])}
+               name={data.original[0]}
+               onChange={this.handleCheckboxClick} />);
+    }
+
+    renderSelectAllCheckbox () {
+        // renders a select all checkbox, sets the check to
+        // indeterminate if only some of the visible resources
+        // are checked
+        let {all, checked}  = this.selected;
```
Comment (adngdb) [inline pontoon/static/js/tags/admin/table.js]: Double white space. :)

File: pontoon/static/js/widgets/errors.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+
+import React from 'react';
+
+
+export default class ErrorList extends React.Component {
```
Comment (adngdb) [inline pontoon/static/js/widgets/errors.js]: I've seen several code examples where people use functions for components that don't need lifecycle stuff. This would be a good fit. What do you think of that?

File: pontoon/static/js/widgets/errors.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+    render () {
+        const {errors} = this.props;
+        if (!Object.keys(errors).length > 0) {
```
Comment (adngdb) [inline pontoon/static/js/widgets/errors.js]: This is confusing. Why not just `length === 0`? Or `!length`?

File: tests/js/tags/admin.js [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+           tag='foo'
+           api='bar'
+           project='baz' />);
+    let result = button.instance().renderResourceManager();
+    expect(result).toBe('');
+    button.setState({open: true});
+    result = button.instance().renderResourceManager();
+    expect(result.props.tag).toBe('foo');
+    expect(result.props.api).toBe('bar');
+    expect(result.props.project).toBe('baz');
```
Comment (adngdb) [inline tests/js/tags/admin.js]: This puzzles me. Is this not just testing React? I think we can, and probably should, live without tests like that. They increase the size of the code base and the maintenance burden for very little value.

File: pontoon/static/js/tags/admin/index.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+               api={node.dataset.api}
+               tag={node.dataset.tag}
+               project={node.dataset.project} />,
+            node);
+    }
+
+    addButtons () {
+        this.tagResourceNodes.forEach(this.addButton);
+    }
+}
```
[the remark this answers is not in the document]
Comment (phlax) [reply pontoon/static/js/tags/admin/index.js]: > I don't understand why this is a class.

I can very much imagine this being overridden as a generic loader - i would defo tend towards classes over functions - even with js

File: pontoon/static/js/tags/admin/table.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+            all = true;
+        }
+        return {all: all, checked: checked};
+    }
+
+    componentWillUpdate (prevProps) {
+        if (prevProps.data !== this.props.data) {
+            const {checked} = this.state;
+            this.visible.length = 0;
+            checked.clear();
```
[the remark this answers is not in the document]
Comment (phlax) [reply pontoon/static/js/tags/admin/table.js]: intentionally so

if you call setState in all of these places it recurses - simply emptying the set works

File: tests/js/tags/admin.js (hunk 6 above)
Comment (phlax) [reply tests/js/tags/admin.js]: > This puzzles me. Is this not just testing React? 

no, definitely not - its testing that our component is taking its vars from the html element

File: pontoon/static/js/widgets/errors.js (hunk 5 above)
Comment (phlax) [reply pontoon/static/js/widgets/errors.js]: ill update

File: pontoon/static/js/widgets/errors.js (hunk 4 above)
Comment (phlax) [reply pontoon/static/js/widgets/errors.js]: sounds like a good idea, but i would need to investigate or be shown an example to be sure i follow

File: pontoon/static/js/tags/admin/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (adngdb) [reply pontoon/static/js/tags/admin/index.js]: I get your point, but I'll argue that this is increasing complexity for the sake of maybe making something simpler in the future. My worry is that we might never actually need a loader, and we thus simply have code that's more complex for no good reason. I've seen this happen on Socorro, and make some parts of the code base a nightmare to maintain, because the code was way too generic. I'd love to avoid this happening on Pontoon. :)

File: pontoon/static/js/tags/admin/table.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (adngdb) [reply pontoon/static/js/tags/admin/table.js]: Well doesn't that imply there's an underlying architectural problem then? Isn't there another way of writing this so that we can use `setState` to change the state everywhere?

File: pontoon/static/js/widgets/errors.js (hunk 4 above)
Comment (adngdb) [reply pontoon/static/js/widgets/errors.js]: That'd look like 
```javascript
const ErrorList = ({ errors }) => {
    // This is basically the render() function.
    return <ul></ul>
};
export default ErrorList;
```

It's being used in the React docs, like here: https://reactjs.org/docs/components-and-props.html

File: tests/js/tags/admin.js (hunk 6 above)
Comment (adngdb) [reply tests/js/tags/admin.js]: Ah, OK, I missed that there's a level of abstraction here. My bad! :)

File: pontoon/static/js/tags/admin/table.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (phlax) [reply pontoon/static/js/tags/admin/table.js]: i dont see that tbh - i think what is happening here is we are saying - we want to empty (etc) that set, but we dont want to force a refresh

[the remark this answers is not in the document]
Comment (phlax) [reply pontoon/static/js/tags/admin/table.js]: ok, i went through code and found that it was either already setting state or was ok to wherever needed - ive updated so that it never touches the copy of the checked set itself

File: pontoon/static/js/tags/admin/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (phlax) [reply pontoon/static/js/tags/admin/index.js]: Im going to have to disagree here. Well structured classes, with good names for methods and properties are more readable, maintainable, extensible and testable than procedural code with closure, im struggling to see how you think the example you gave is easier to understand

File: pontoon/static/js/widgets/errors.js (hunk 4 above)
Comment (phlax) [reply pontoon/static/js/widgets/errors.js]: im back to the why use a function question. Functional programming feels to me like a javascript throwback to a time when it didnt have the basics of object-oriented programming. It  also makes it difficult to extend the errorlist in any way - something i can certainly imagine doing

File: pontoon/static/js/tags/admin/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (Pike) [reply pontoon/static/js/tags/admin/index.js]: As a general comment, if there's code in need of more detail to understand, let's add comments, and not more code.

Also, in this particular example, there's nothing to test in each method, aside of the Array.from and NodeList implementations in browsers. the only way to test that this code actually does what it should is, at this level, an integration test of the function or class.

And please, no generic base class, in a separate file, even, for use-cases like this.

Nit on Adrian's code, there's an optional second argument to `Array.from`, which comes in handy in cases like this. You can pass in a function that'll be called for each item before the Array is constructed. One-liner.

File: pontoon/static/js/widgets/errors.js (hunk 4 above)
Comment (phlax) [reply pontoon/static/js/widgets/errors.js]: Ive switched to using a `PureComponent` which i believe provides the same benefits as a stateless function, but with a class structure

File: pontoon/static/js/tags/admin/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (phlax) [reply pontoon/static/js/tags/admin/index.js]: > As a general comment, if there's code in need of more detail to understand, let's add comments, and not more code

For me, its not so much about adding more/less code, its more about preferring an object-oriented style and use of data structures over functional/closure style.

> Also, in this particular example, there's nothing to test in each method

I disagree. We want to know that the selectors being used in the js widget match the ones being used when generating the templates:

https://github.com/phlax/pontoon/blob/658bd64a17eee5d779c6a226070184974e2d993b/tests/js/tags/admin.js#L486

There is literally nothing else that ties these 2 disparate parts of the code together.

> And please, no generic base class, in a separate file, even, for use-cases like this.

None were used. My point was more that the thing that has sold me on React, is the ease with which i can build components, see common patterns, factor out the useful stuff, rinse and repeat.

[the remark this answers is not in the document]
Comment (adngdb) [reply pontoon/static/js/tags/admin/index.js]: There are three things that bother me, and make me favor "my way". First, I believe code complexity is highly coupled to the number of lines of code. Less lines of code, means less things to read, which, most of the time, means code is simpler to understand. 

Second, in this specific case, using a class means that to understand how the algorithm plays out, you have to jump from method to method. With the function, it goes from top to bottom, you just have to read it straight, no "goto". It seems a lot simpler to me, especially with such a small function. Why do complicated if you can do simple? (as we often say in French, not sure that is an actual idiom in English... :P ) That leads me to believe this is a case of premature optimization. You write code anticipating use cases that do not exist yet, and might never exist. But I've made that point already.

My last point goes deeper, it's on using OOP. I've grown very wary of OOP, because it leads to some antipatterns that can be very difficult to deal with, especially in big code bases. The more I code, and the more I use the React + redux environment, the more I think OOP should be the exception. I do use classes when it makes sense, for stateful components for example. But I avoid inheritance as much as possible, favoring composition instead. 

The best resources I have on this point are these two articles by Eric Elliott, on what he calls the "Two pillars of JS": https://medium.com/javascript-scene/the-two-pillars-of-javascript-ee6f3281e7f3 and https://medium.com/javascript-scene/the-two-pillars-of-javascript-pt-2-functional-programming-a63aa53a41a4

[the remark this answers is not in the document]
Comment (phlax) [reply pontoon/static/js/tags/admin/index.js]: hmm - i definitely get that functional programming has classically been the style of javascript - and could find many articles bestowing its virtue (and tbh critiquing it).

imhe, its difficult to test, and leads to heavily nested code in which its very difficult to keep track of local vars. I have personally debugged a lot of javascript code in which these vars were for one reason or another not being garbage collected. Found using a profiler, because the code was so hard to follow.

I think modern javascript has to a large extent addressed these issues by:

- introducing a more familiar OOP pattern (i get that this is "syntactic sugar" but its powerful all the same)
- introducing (like python) `await` syntax that lets you not have to keep embedding functionality in ever more nested code

just my 2p, but i would like to take adavantage of these advances

[the remark this answers is not in the document]
Comment (Pike) [reply pontoon/static/js/tags/admin/index.js]: After studying the tests, I've opened an email thread about the goals of our testing. In this particular case, yes, the tests only test the additional lines of code around the actual for loop, and validate that they don't break. I'd make that point on the actual tests, but github refuses to load that part of the diff for me.

React-newbie question: where are the <div class="js-tag-resources"> elements actually created? Didn't find that.

[the remark this answers is not in the document]
Comment (phlax) [reply pontoon/static/js/tags/admin/index.js]: @Pike in the following PR here - #802

[the remark this answers is not in the document]
Comment (phlax) [reply pontoon/static/js/tags/admin/index.js]: more specifically here https://github.com/mozilla/pontoon/pull/802/files#diff-d7b0530da8eb2ab60bc18f6d157c4a73R24

File: pontoon/static/js/tags/admin/table.js [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        this.setState({checked});
+    }
+
+    handleSelectAll () {
+        // user clicked the select all checkbox...
+        // if there are some resources checked already, all are removed
+        // otherwise all visible are checked.
+        let {checked} = this.state;
+        checked = checked.size > 0 ? new Set() : new Set([...this.visible])
+        return this.setState({checked})
```
Comment (adngdb) [inline pontoon/static/js/tags/admin/table.js]: No need to return here, right?

File: pontoon/static/js/tags/admin/table.js [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    async handleSubmit (evt) {
+        // after emitting handleSubmit to parent with list of currently
+        // checked, clears the checkboxes
+        evt.preventDefault();
+        const {checked} = this.state;
+        await this.props.handleSubmit(Array.from(checked));
```
Comment (adngdb) [inline pontoon/static/js/tags/admin/table.js]: Is there a difference between `[...checked]` and `Array.from(checked)`? I see you're using both, so I wonder if we should maybe choose one and stick to it. As far as I can tell, they do the same thing, right?

Comment (phlax) [reply pontoon/static/js/tags/admin/table.js]: yep, i think you are right - personally i prefer spread destructuring

File: pontoon/static/js/tags/admin/table.js (hunk 9 above)
Comment (phlax) [reply pontoon/static/js/tags/admin/table.js]: i think there is - at least for testing purposes - setState is async so without this you cant check that state has changed

Comment (adngdb) [reply pontoon/static/js/tags/admin/table.js]: I don't get it. Using `return` makes the function wait of asynchronous calls? Also, `setState` doesn't return anything.

Comment (phlax) [reply pontoon/static/js/tags/admin/table.js]: > return makes the function wait of asynchronous calls? 

yep, i think here it does - and certainly if you want to wait on it in tests

> setState doesn't return anything.

it returns a promise i think (except in componentdidmount iirc)

File: tests/js/tags/admin.js [hunk 11]
```diff
@@ -0,0 +1,498 @@
...
+
+
+test('TagResourcesLoader constructor', () => {
+    document.querySelectorAll = jest.fn(() => new Set([1, 2, 3]));
+    const loader = new TagResourcesLoader();
+    const nodes = loader.tagResourceNodes;
+    expect(nodes).toEqual([1, 2, 3]);
+    expect(document.querySelectorAll.mock.calls).toEqual(
+        [["div.js-tag-resources"]]);
+})
```
Comment (Pike) [inline tests/js/tags/admin.js]: I've got good comments on the patterns here below on the test on 477ff. Sorry for the reverse-polish notation

File: tests/js/tags/admin.js [hunk 12]
```diff
@@ -0,0 +1,498 @@
...
+test('TagResourcesLoader addButtons', () => {
+    document.querySelectorAll = jest.fn(() => ['a', 'b', 'c']);
+    const loader = new TagResourcesLoader();
+    loader.addButton = jest.fn();
+    loader.addButtons();
+    // forEach on [a, b, c]
+    expect(loader.addButton.mock.calls).toEqual(
+        [["a", 0, ["a", "b", "c"]],
+         ["b", 1, ["a", "b", "c"]],
+         ["c", 2, ["a", "b", "c"]]]);
```
[the remark this answers is not in the document]
Comment (phlax) [reply tests/js/tags/admin.js]: no its testing that addButton was called

[the remark this answers is not in the document]
Comment (phlax) [reply tests/js/tags/admin.js]: and with what args

File: tests/js/tags/admin.js [hunk 13]
```diff
@@ -0,0 +1,498 @@
...
+    expect(document.querySelectorAll.mock.calls).toEqual(
+        [["div.js-tag-resources"]]);
```
[the remark this answers is not in the document]
Comment (phlax) [reply tests/js/tags/admin.js]: its testing that that querySelectorAll was called with the target selector

File: tests/js/tags/admin.js [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+})
+
+
+test('TagResourcesLoader loaded', () => {
+    TagResourcesLoader.prototype.addButtons = jest.fn()
+    const DOMContentLoaded_event = document.createEvent("Event")
+    DOMContentLoaded_event.initEvent("DOMContentLoaded", true, true)
+    window.document.dispatchEvent(DOMContentLoaded_event)
```
[the remark this answers is not in the document]
Comment (phlax) [reply tests/js/tags/admin.js]: its testing that *when* domcontentloaded event is fired, the loader is loaded

File: tests/js/tags/admin.js [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+
+
+test('TagResourcesLoader addButton', () => {
+    const loader = new TagResourcesLoader();
+    ReactDOM.render = jest.fn();
+    const button = {dataset: {api: 'API0', project: 'PROJECT0', tag: 'TAG0'}};
+    loader.addButton(button);
+    const tagButton = ReactDOM.render.mock.calls[0][0];
+    expect(tagButton.props).toEqual(button.dataset);
+    expect(ReactDOM.render.mock.calls[0][1]).toEqual(button);
```
[the remark this answers is not in the document]
Comment (phlax) [reply tests/js/tags/admin.js]: its testing that ReactDOM.render gets called with the right things

[the remark this answers is not in the document]
Comment (phlax) [reply tests/js/tags/admin.js]: im wondering if there is some misunderstanding here.

`ReactDOM.render` does not test the rendering of a `component` as such, it tests the *loading* of a component into the `DOM`

File: tests/js/tags/admin.js [hunk 16]
```diff
@@ -0,0 +1,498 @@
...
+    expect(manager.state()).toEqual(
+        {type: 'assoc', search: 'FOO', data: [], errors: {}});
+    evt.target.nodeName = 'SELECT';
+    evt.target.value = 'BAR';
+    result = await manager.instance().handleSearchChange(evt);
+    expect(result).toBe(23);
+    expect(_refreshData.mock.calls).toEqual([[], [], []]);
+    expect(manager.state()).toEqual(
+        {type: 'BAR', search: 'FOO', data: [], errors: {}});
+});
```
[the remark this answers is not in the document]
Comment (phlax) [reply tests/js/tags/admin.js]: @adrian we can use whatever we want here - i use method names because they are unit tests and you mostly want to check what its testing against.

[the remark this answers is not in the document]
Comment (phlax) [reply tests/js/tags/admin.js]: ...ill check this test tho, and see if i can make it clearer

File: tests/js/tags/admin.js [hunk 17]
```diff
@@ -0,0 +1,498 @@
...
+            _refreshData();
+            return 23;
+        }
+    }
+
+    const manager = shallow(<MockTagResourceManager />);
+    const evt = {target: {nodeName: 'INPUT', value: "FOO"}};
+    let result = await manager.instance().handleSearchChange(evt);
+    expect(result).toBe(23);
+    expect(_refreshData.mock.calls).toEqual([[], []]);
```
[the remark this answers is not in the document]
Comment (phlax) [reply tests/js/tags/admin.js]: its testing what its been called with, but may be superfluous

File: tests/js/tags/admin.js [hunk 18]
```diff
@@ -0,0 +1,498 @@
...
+
+        async refreshData () {
+            _refreshData();
+        }
+    }
+
+    const manager = shallow(<MockTagResourceManager />);
+    expect(manager.text()).toBe("<TagResourceSearch /><ErrorList /><TagResourceTable />");
+    expect(_refreshData.mock.calls).toEqual([[]]);
+});
```
[the remark this answers is not in the document]
Comment (phlax) [reply tests/js/tags/admin.js]: > I am also very unsure about testing the output of the render method. In django, it would like testing that a view returns the correct HTML. I can see myself doing that, but only for very specific cases where the markup is super important.

mostly its doing `shallow` renders - this is the accepted way to test component rendering without actually rendering, i believe.

in some cases, it wants to test specific functionality - and these  tests use `mount` rather than `shallow`

File: tests/js/tags/admin.js [hunk 19]
```diff
@@ -0,0 +1,498 @@
...
+           api='bar'
+           project='baz' />);
+    let result = button.instance().renderResourceManager();
+    expect(result).toBe('');
+    button.setState({open: true});
+    result = button.instance().renderResourceManager();
+    expect(result.props.tag).toBe('foo');
+    expect(result.props.api).toBe('bar');
+    expect(result.props.project).toBe('baz');
+});
```
[the remark this answers is not in the document]
Comment (phlax) [reply tests/js/tags/admin.js]: this is a really important test as the widget depends on these vars being passed through

**Comment by phlax** [comment]
> I guess I can summarize that by saying that these tests are hard to read, because they are focused on testing implementation details rather than "stories" (i.e. use cases, user stories, test cases... ). I don't think what they're doing is necessarily wrong, my concern is that it will be hard, a year from now, to go back to those tests, understand them and edit them. I believe organizing tests around those so-called "stories" will help mitigate that problem, and make them a lot easier to maintain.

i get the point about being hard to read, i can add some comments to make clearer the *intention* and not just the *implementation*.

On the point about narratives, these are (mostly) unit tests so they by definition dont express or test a narrative, they test a unit of computation or processing

**Comment by Pike** [comment]
My diff for the loader is on https://github.com/phlax/pontoon/compare/string_tiers_admin_js...Pike:string_tiers_admin_js. It creates a named function `onScriptLoad` to actually test the `addEventListener`, and it also names the function to check for.

That function is also flat, and tests querySelectorAll, as well as argument-propagation through to `.props` and the right node. Also note the choice of `toBe` vs `toEqual` in the tests there.

**Comment by phlax** [comment]
thanks @Pike ill take a look asap, busy with the queryset stuff atm

**Comment by phlax** [comment]
@adngdb ...

> However, please please please do not rebase/squash your commits until we've been through the entire review 

massive apology, its a force of habit thing - ive been trying hard not to

>  Too "unit" tests, in a way, ha ha... I guess I favor something that is a bit higher-level,

defo not against that, but from a land-code-quickly-and-tested pov unit test are a good starting point i think

> I'm going with class components all the time now. 

big smiles from /me !!!

> However I still believe TagResourcesLoader shouldn't be a component at all, and just a function, as I initially stated. I assume that's the point @Pike is going to provide a diff for?

in this particular case im more than happy to compromise and will look at patch asap. My overriding thought however is that this pattern - ie `find nodes` > `build components` > `insert to DOM` is one that we are going to visit a lot and could well be served by a generic `class`. For this PR however i would much rather we focus on the logic (and bugs 8/ ) here

File: tests/js/tags/admin.js [hunk 20]
```diff
@@ -0,0 +1,848 @@
...
+            _refreshData();
+        }
+    }
+
+    const manager = shallow(<MockTagResourceManager />);
+
+    // The TagResourceManager has 3 sub components
+    expect(manager.text()).toBe("<TagResourceSearch /><ErrorList /><TagResourceTable />");
+
+    // enzyme.shallow triggers `componentDidUpdate` - so we expect
```
Comment (phlax) [inline tests/js/tags/admin.js]: this should say `componentDidMount`

File: tests/js/tags/admin.js [hunk 21]
```diff
@@ -0,0 +1,848 @@
...
+
+    // it has been called twice now, because it was called when
+    // component was (shallow) rendered (with no args).
+    expect(_refreshData.mock.calls).toEqual([[], []]);
+
+    // as our mock event was text input, we have updated the
+    // search string in `state`
+    expect(manager.state()).toEqual(
+        {type: 'assoc', search: 'FOO', data: [], errors: {}});
+    evt.target.nodeName = 'SELECT';
```
Comment (phlax) [inline tests/js/tags/admin.js]: move these lines below comment

File: tests/js/tags/admin.js [hunk 22]
```diff
@@ -0,0 +1,848 @@
...
+    class MockTagResourceManager extends TagResourceManager {
+
+        async refreshData () {
+            _refreshData();
+        }
+    }
+
+    const manager = shallow(<MockTagResourceManager />);
+    manager.instance().setState = jest.fn(async () => 23);
+    const response = {status: 500, json: jest.fn(async () => ({}))};
```
Comment (phlax) [inline tests/js/tags/admin.js]: move this line below comment

File: pontoon/static/js/tags/admin/table.js (hunk 9 above)
Comment (phlax) [reply pontoon/static/js/tags/admin/table.js]: yep im pretty confident that a) setState is async and b) you cant test against it if you dont handle the promises.

There is quite a lot of stuff on the interwebs about the async nature of setState - this seems to sum it fairly well 
 https://stackoverflow.com/questions/42038590/when-to-use-react-setstate-callback 

this also has some interesting discussion:

https://medium.com/front-end-hacking/async-await-with-react-lifecycle-methods-802e7760d802

im pretty sure that it works differently when called in `componentDidMount` but struggling to find hard evidence of that

File: pontoon/static/js/tags/admin/table.js [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+               onClick={this.handleSubmit}>
+              {this.submitMessage}
+            </button>);
+    }
+
+    renderTable () {
+        return (
+            <ReactTable
+               defaultPageSize={this.defaultPageSize}
+               className="-striped -highlight"
```
[the remark this answers is not in the document]
Comment (phlax) [reply pontoon/static/js/tags/admin/table.js]: its part of the `ReactTable` api - see here https://react-table.js.org/#styles

**Comment by phlax** [comment]
Closing - all code from here has either landed or is in #802
